### PR TITLE
Follow up items for PR 1159

### PR DESCRIPTION
**Pull request by m-nash**
- Follow up items for https://github.com/Azure/autorest.csharp/pull/1159
- Adds a new SystemObjectType which inherits from ObjectType but wraps a System.Type so that it can be used in all writers seemlessly
- Moved majority of implementation for ObjectType into SchemaObjectType and made ObjectType mostly abstract implementation to support the SystemObjectType
- Eliminated the need for a special ModelWriter and SerializationWriter in mgmt plane since we can now use the SystemObjectType

### CODE REVIEW

File: src/AutoRest.CSharp/Mgmt/Decorator/InheritanceChoser.cs [hunk 1]
```diff
@@ -213,5 +213,16 @@ private static bool IsSuperset(MgmtObjectType childType, System.Type parentType)
             }
             return parentProperties.Count == matchCount;
         }
+
+        private class Node
```
Comment (chamons) [inline src/AutoRest.CSharp/Mgmt/Decorator/InheritanceChoser.cs:217]: Maybe use a record class here?

**Comment by chamons** [comment]
As Pavel did the last review I'll let him give the 👍

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Linq;
+using System.Reflection;
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs:6]: Unnecessary diff here?

**Comment by chamons** [comment]
This is turning out to be a large change, I'd like time for Pavel and I to review it before it goes in.

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs [hunk 3]
```diff
@@ -568,6 +572,11 @@ private static void DeserializeFrameworkTypeValue(CodeWriter writer, CodeWriterD
             writer.AppendRaw(")");
         }
 
+        private static bool IsMgmtReferenceType(Type frameworkType)
```
Comment (allenjzhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:575]: nit, as this list may change, would like to keep this code in Mgmt rather than Common. Can we call out to a static checked to there?

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs [hunk 4]
```diff
@@ -1,7 +1,6 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-using System;
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs:4]: Strange diff to include.

File: src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs [hunk 5]
```diff
@@ -2,129 +2,88 @@
...
+
+            if (!IncludeDeserializer)
+            {
+                yield break;
+            }
+
+            // Skip serialization ctor if they are the same
+            if (!InitializationConstructor.Parameters
+                .Select(p => p.Type)
+                .SequenceEqual(SerializationConstructor!.Parameters.Select(p => p.Type)))
```
Comment (chamons) [inline src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs:91]: This used to be:
-                .SequenceEqual(SerializationConstructor.Parameters.Select(p => p.Type)))
+                .SequenceEqual(SerializationConstructor!.Parameters.Select(p => p.Type)))

Could SerializationConstructor be null here?

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                }
+            }
+
+            return new List<ParameterInfo>();
+        }
+
+        internal IEnumerable<PropertyInfo> GetAllProperties()
+        {
+            var type = _type;
+            while (type is not null)
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:65]: Same as above.

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                foreach (var property in type.GetProperties())
+                {
+                    yield return property;
+                }
+                type = type.BaseType;
+            }
+        }
+
+        protected override ObjectTypeDiscriminator? BuildDiscriminator()
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:75]: `protected override ObjectTypeDiscriminator? BuildDiscriminator() => null` ?

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        protected override IEnumerable<ObjectTypeProperty> BuildProperties()
+        {
+            foreach (var property in _myProperties)
+            {
+                var getter = property.GetGetMethod();
+                var setter = property.GetSetMethod();
+                MemberDeclarationOptions memberDeclarationOptions = new MemberDeclarationOptions(
+                    getter is not null && getter.IsPublic ? "public" : "internal",
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:118]: Again.

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+                }
+                yield return new ObjectTypeProperty(memberDeclarationOptions, prop.Summary, prop.IsReadOnly, prop, GetTypeFromSystem(property.PropertyType));
+            }
+        }
+
+        private bool GetReadOnly(PropertyInfo property)
+        {
+            if (property.Name == "Tags")
+                return false;
+            return property.GetSetMethod() is null;
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:140]: Again.

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                {
+                    args.Add(GetTypeFromSystem(argType));
+                }
+            }
+            return new CSharpType(type, false, args.ToArray());
+        }
+
+        private string GetPropertySummary(MethodInfo? setter)
+        {
+            return setter is not null ? " or sets" : string.Empty;
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:158]:
```suggestion
            return setter != null ? " or sets" : string.Empty;
```

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        protected override CSharpType? CreateInheritedDictionaryType()
+        {
+            throw new NotImplementedException();
+        }
+
+        protected override CSharpType? CreateInheritedType()
+        {
+            return _type.BaseType is null ? null : CSharpType.FromSystemType(Context, _type.BaseType); //new CSharpType(_type.BaseType);
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:178]:
```suggestion
            return _type.BaseType == null ? null : CSharpType.FromSystemType(Context, _type.BaseType); //new CSharpType(_type.BaseType);
```

**Review by chamons — CHANGES_REQUESTED** [review]
This needs @pakrym 👀 as it touched deep type system types.

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+                    getter is not null && getter.IsPublic ? "public" : "internal",
+                    property.Name,
+                    GetTypeFromSystem(property.PropertyType));
+                Property prop = new Property();
+                prop.Nullable = false;
+                prop.ReadOnly = GetReadOnly(property); //TODO read this from attribute from reference object
+                prop.SerializedName = ToCamelCase(property.Name);
+                prop.Summary = $"Gets{GetPropertySummary(setter)} {property.Name}";
+                prop.Required = true;
+                if (property.PropertyType.IsGenericType && property.PropertyType.GetGenericTypeDefinition() == typeof(IDictionary<,>))
```
Comment (allenjzhang) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:127]: Can we add comment here to capture the node on the just enough type supported for Mgmt.

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs [hunk 13]
```diff
@@ -37,13 +37,15 @@ public CSharpType(Type type, bool isNullable, params CSharpType[] arguments)
...
 
-        public CSharpType(TypeProvider implementation, string ns, string name, bool isValueType = false, bool isNullable = false)
+        public CSharpType(TypeProvider implementation, string ns, string name, bool isValueType = false, bool isNullable = false, CSharpType[]? arguments = default)
         {
             _implementation = implementation;
             Name = name;
             IsValueType = isValueType;
             IsNullable = isNullable;
             Namespace = ns;
+            if (arguments is not null)
```
[the remark this answers is not in the document]
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:47]: is not null is new syntax that avoids the operator overload null ref exception.  same as "is null" but they finally added a not version of it.

The property is initialized to an empty set if its never used so this maintains that default behavior

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs [hunk 14]
```diff
@@ -87,5 +89,18 @@ public override string ToString()
         {
             return new CodeWriter().Append($"{this}").ToString(false);
         }
+
+        internal static CSharpType FromSystemType(BuildContext context, Type parentType)
+        {
+            var genericTypes = parentType.GetGenericArguments().Select<Type, CSharpType>(t => new CSharpType(t));
```
[the remark this answers is not in the document]
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs]: to use var yes, if i had an explicit variable type then maybe it could determine it

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs (hunk 4 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs:4]: I just did a remove all unused usings in a bunch of files i was touching.  Happy to undo this if you want

File: src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs (hunk 5 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs:91]: Its null for the SystemObjectType path and not null in the SchemaObjectType path.  So in this class its safe that it won't be null.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs (hunk 4 above)
Comment (chamons) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeConstructor.cs:4]: That's a fine reason.

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:47]: I'm not a fan of `is not null` or `is null` because while I understand the theoretical benefit I'm yet to see the problem in practice.

[the remark this answers is not in the document]
Comment (pakrym) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:47]: Consider `Arguments = arguments ?? Array.Empty<CSharpType>()`

File: src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs [hunk 15]
```diff
@@ -126,7 +126,8 @@ internal static bool IsDictionary(CSharpType type)
         {
             return type.IsFrameworkType &&
                    (type.FrameworkType == typeof(IDictionary<,>) ||
-                   type.FrameworkType == typeof(IReadOnlyDictionary<,>));
+                   type.FrameworkType == typeof(IReadOnlyDictionary<,>) ||
+                   type.FrameworkType.IsGenericType && type.FrameworkType.GetGenericTypeDefinition() == typeof(IDictionary<,>));
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs:130]: The `type.FrameworkType` should already be the generic definition.

File: src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs [hunk 16]
```diff
@@ -228,7 +228,7 @@ public static CodeWriter WriteReferenceOrConstant(this CodeWriter writer, Refere
             this CodeWriter writer,
             Action<CodeWriter, CodeWriterDelegate> valueCallback,
             ObjectType objectType,
-            ObjectTypeConstructor constructor,
+            ObjectTypeConstructor? constructor,
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs:231]: Why do we need the ability to pass null here? There should be a ctor that we are calling.

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs [hunk 17]
```diff
@@ -159,7 +157,8 @@ public static void ToSerializeCall(this CodeWriter writer, JsonSerialization ser
                         }
                         else if (frameworkType == typeof(string) ||
                                  frameworkType == typeof(char) ||
-                                 frameworkType == typeof(Guid))
+                                 frameworkType == typeof(Guid) ||
+                                 ReferenceTypes.IsMgmtReferenceType(frameworkType))
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:161]: I wonder if we can extend the `JsonValueSerialization` with the concept of "SerializeAs" that would specify the resulting json type...

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs [hunk 18]
```diff
@@ -3,107 +3,72 @@
...
+        public ObjectTypeConstructor? SerializationConstructor => _serializationConstructor ??= BuildSerializationConstructor();
+        public CSharpType? ImplementsDictionaryType => _implementsDictionaryType ??= CreateInheritedDictionaryType();
+        public ObjectTypeConstructor InitializationConstructor => _initializationConstructor ??= BuildInitializationConstructor();
         public ObjectSerialization[] Serializations => _serializations ??= BuildSerializations();
+        public bool IsStruct => ExistingType?.IsValueType == true;
+        public virtual string? Description { get; protected set; }
+        public ObjectTypeDiscriminator? Discriminator => _discriminator ??= BuildDiscriminator();
 
-        public ObjectTypeProperty[] Properties => _properties ??= BuildProperties().ToArray();
+        public abstract bool IncludeSerializer { get; }
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs:41]: I think this makes sense on SchemaObject type only. As well as IncludeDeserializer and IncludeConverter. And possibly AdditionalPropertiesProperty

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs [hunk 19]
```diff
@@ -3,107 +3,72 @@
...
-        protected override string DefaultName => OjectSchema.CSharpName();
-        protected override string DefaultAccessibility { get; } = "public";
-        protected override string DefaultNamespace { get; }
-        protected override TypeKind TypeKind => IsStruct ? TypeKind.Struct : TypeKind.Class;
-
-        public string? Description { get; }
+        public ObjectTypeConstructor[] Constructors => _constructors ??= BuildConstructors().ToArray();
+        public virtual ObjectTypeProperty[] Properties => _properties ??= BuildProperties().ToArray();
 
         public virtual CSharpType? Inherits => _inheritsType ??= CreateInheritedType();
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs:32]: both this property and CreateInheritedType are virtual. One should be enough.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs [hunk 20]
```diff
@@ -3,107 +3,72 @@
...
-
-        public bool IsStruct => ExistingType?.IsValueType == true;
-        protected override string DefaultName => OjectSchema.CSharpName();
-        protected override string DefaultAccessibility { get; } = "public";
-        protected override string DefaultNamespace { get; }
-        protected override TypeKind TypeKind => IsStruct ? TypeKind.Struct : TypeKind.Class;
-
-        public string? Description { get; }
+        public ObjectTypeConstructor[] Constructors => _constructors ??= BuildConstructors().ToArray();
+        public virtual ObjectTypeProperty[] Properties => _properties ??= BuildProperties().ToArray();
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs:30]: Both this and BuildProperties are virtual. We should pick one.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs [hunk 21]
```diff
@@ -3,107 +3,72 @@
...
-        public ObjectTypeProperty[] Properties => _properties ??= BuildProperties().ToArray();
+        public abstract bool IncludeSerializer { get; }
+        public abstract bool IncludeDeserializer { get; }
+        public abstract bool IncludeConverter { get; }
+        public abstract ObjectTypeProperty? AdditionalPropertiesProperty { get; }
+        protected abstract ObjectTypeDiscriminator? BuildDiscriminator();
+        protected abstract ObjectSerialization[] BuildSerializations();
+        protected abstract ObjectTypeConstructor BuildInitializationConstructor();
+        protected abstract CSharpType? CreateInheritedDictionaryType();
+        protected abstract ObjectTypeConstructor? BuildSerializationConstructor();
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs:49]: I'm not a fan of having the mixed extensibility story where some properties are virtual by themselves and other have virtual caching builders. I would pick one.

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+            {
+                foreach (var property in type.GetProperties())
+                {
+                    yield return property;
+                }
+                type = type.BaseType;
+            }
+        }
+
+        protected override ObjectTypeDiscriminator? BuildDiscriminator() => null;
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:75]: Do we need the disctiminator on the ObjectType then?

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+        public override bool IncludeConverter => false;
+
+        public override string DefaultName => GetNameWithoutGeneric(_type);
+
+        private string ToCamelCase(string name)
+        {
+            return name.Substring(0, 1).ToLower(CultureInfo.InvariantCulture) + name.Substring(1);
+        }
+
+        private IEnumerable<PropertyInfo> _myProperties => _type.GetProperties().Where(p => p.DeclaringType == _type);
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:47]: Inline?

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+            }
+            return new CSharpType(type, false, args.ToArray());
+        }
+
+        private string GetPropertySummary(MethodInfo? setter)
+        {
+            return setter is not null ? " or sets" : string.Empty;
+        }
+
+        protected override ObjectTypeConstructor? BuildSerializationConstructor() => null;
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:163]: I have a suspicion that everything that's not implemented on the SystemObjectType can be moved from ObjectType to SchemaObjectType

File: src/AutoRest.CSharp/Common/Output/Models/Types/TypeProvider.cs [hunk 25]
```diff
@@ -27,10 +28,10 @@ protected TypeProvider(BuildContext context)
             TypeKind == TypeKind.Struct || TypeKind == TypeKind.Enum);
         public TypeDeclarationOptions Declaration => _type ??= BuildType();
 
-        protected BuildContext Context { get; private set; }
-        protected abstract string DefaultName { get; }
+        internal BuildContext Context { get; private set; }
+        public abstract string DefaultName { get; }
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/TypeProvider.cs:32]: We shouldn't use this publicly. As sometimes it's redefined.

File: src/AutoRest.CSharp/DataPlane/Output/DataPlaneClient.cs [hunk 26]
```diff
@@ -29,6 +29,7 @@ public DataPlaneClient(OperationGroup operationGroup, BuildContext<DataPlaneOutp
         {
             _operationGroup = operationGroup;
             _context = context;
+            DefaultAccessibility = "public";
```
Comment (pakrym) [inline src/AutoRest.CSharp/DataPlane/Output/DataPlaneClient.cs:32]: public is the default

File: src/AutoRest.CSharp/DataPlane/Output/DataPlaneResponseHeaderGroupType.cs [hunk 27]
```diff
@@ -22,6 +22,7 @@ internal class DataPlaneResponseHeaderGroupType: TypeProvider
 
         public DataPlaneResponseHeaderGroupType(OperationGroup operationGroup, Operation operation, HttpResponseHeader[] httpResponseHeaders, BuildContext<DataPlaneOutputLibrary> context) : base(context)
         {
+            DefaultAccessibility = "internal";
```
Comment (pakrym) [inline src/AutoRest.CSharp/DataPlane/Output/DataPlaneResponseHeaderGroupType.cs:25]: nit: move next to the default name assignment

File: src/AutoRest.CSharp/Common/Output/Models/Types/TypeProvider.cs [hunk 28]
```diff
@@ -18,6 +18,7 @@ protected TypeProvider(BuildContext context)
         {
             Context = context;
             _existingType = new Lazy<INamedTypeSymbol?>(() => Context.SourceInputModel?.FindForType(DefaultNamespace, DefaultName));
+            DefaultAccessibility = "public";
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/Types/TypeProvider.cs:21]: What's the benefit of this change?

File: src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs (hunk 16 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs:231]: no longer needed

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs (hunk 17 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:161]: yeah, was thinking the same thing, but given how large this change is already Id suggest we take that up in a later PR

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs (hunk 18 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs:41]: The problem is the value is used in places where its expecting an ObjectType which is why we need the property here.  Otherwise we have to cast in dozens of places.  Having it here allows us to treat SystemObjectType and SchemaObjectType identically from all the writers perspectives

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs (hunk 21 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs:50]: not sure I follow

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs (hunk 22 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:90]: yes because its used by writers which expect an ObjectType

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs (hunk 23 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:49]: sure, i was originally using it in multiple places

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs (hunk 24 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs]: this has an implementation now

File: src/AutoRest.CSharp/Common/Output/Models/Types/TypeProvider.cs (hunk 28 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/TypeProvider.cs:21]: none, originally it was needed for something, but now its not ill revert

File: src/AutoRest.CSharp/DataPlane/Output/DataPlaneResponseHeaderGroupType.cs (hunk 27 above)
Comment (m-nash) [reply src/AutoRest.CSharp/DataPlane/Output/DataPlaneResponseHeaderGroupType.cs:25]: obsolete given comment below

File: src/AutoRest.CSharp/DataPlane/Output/DataPlaneClient.cs (hunk 26 above)
Comment (m-nash) [reply src/AutoRest.CSharp/DataPlane/Output/DataPlaneClient.cs:32]: obsolete given comment below

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs (hunk 20 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs:30]: this one is tricky due to the infinite loop.  Its easier if I can explain by showing you tomorrow.

File: src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs (hunk 15 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs:130]: In the case of SystemObjectType it is not

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (chamons) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:47]: https://xkcd.com/285/ on it being a best practice.

I'm unable to find that listed on MSDN or from dotnet core after a brief search.

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (chamons) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs]: I checked out your branch and was able to remove it without any issues:

https://gist.github.com/chamons/1da9e7258f7a28f57ba13291a1400b26

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs (hunk 17 above)
Comment (chamons) [reply src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:161]: Please file a tracking issue and i'm fine w\ punting this bit.

File: src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs (hunk 5 above)
Comment (chamons) [reply src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs]: I really do not like this "it can be null in one case and not null in another", it is really easy to trip in a future PR. However it is in Mgmt code, so I'm not blocking here.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs (hunk 21 above)
Comment (chamons) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs]: To restate, why do we need both these:

`public ObjectTypeConstructor SerializationConstructor => _serializationConstructor ??= BuildSerializationConstructor();`
And
`protected abstract ObjectTypeConstructor BuildSerializationConstructor();`

and then later have

`public abstract bool IncludeDeserializer { get; }`

Some things are properties you need to define values from, others are methods that the base class uses to define properties for.

If we made `ObjectTypeConstructor SerializationConstructor { get; }` abstract then it would be consistent is the point I believe.

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs (hunk 24 above)
Comment (chamons) [reply src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs]: The implementation is to return empty array. I still don't see why it couldn't move to SchemaObjectType?

File: src/AutoRest.CSharp/DataPlane/Output/DataPlaneClient.cs (hunk 26 above)
Comment (chamons) [reply src/AutoRest.CSharp/DataPlane/Output/DataPlaneClient.cs]: I guess each client has to define accessibility now.

**Review by chamons — CHANGES_REQUESTED** [review]
Work is promising, but it needs some more hammering before going in.

Major
------
- ObjectType.cs double virtual methods and "an infinite loop" explaination
- is not null/is null vs != null/==null
- File tracking issue for JsonCodeWriterExtensions extensibility and mgmt types
- Wait for pavel to respond to 4 other comments

Minor
-------
- Unnecessary generic constraits on Select call
- Builders vs caching abstract methods in ObjectType.cs
- SystemObjecType BuildSerializationConstructor on base type?

Not blocking
-------------------
- Nullability concerns in MgmtObjectType.cs

File: src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs (hunk 15 above)
Comment (pakrym) [reply src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs:130]: then we should make it so. That's the contract of CSharpType.

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs (hunk 22 above)
Comment (pakrym) [reply src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:78]: Why do they expect an ObjectType and not the SchemaType?

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs (hunk 24 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs]: it doesn't return an empty array, and why the ObjectType needs the definition is explained here https://github.com/Azure/autorest.csharp/pull/1170#discussion_r618908466

File: src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs (hunk 22 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs:78]: because there was no objecttype originally, and the whole point of this change was to allow the common writers to work with both schemaobjecttype (the old objecttype) and systemobjecttype in the exact same manner

File: src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs (hunk 21 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs]: that is how it was before.  I am not introducing this its a copy paste from the old objecttype

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs (hunk 17 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:161]: https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_workitems/edit/5812

File: src/AutoRest.CSharp/Mgmt/Decorator/InheritanceChoser.cs (hunk 1 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Mgmt/Decorator/InheritanceChoser.cs:215]: prefer to leave this as is for now, it works and its isolated in mgmt library

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:47]: Ill revert these as it seems you are very passionate about it.

File: src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs (hunk 5 above)
Comment (m-nash) [reply src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs]: It is no longer null based on some changes I made in another area

**Review by pakrym — APPROVED** [review]

**Review by chamons — APPROVED** [review]
